pkg/util/kube/predicates.go

```diff
@@ -108,25 +108,31 @@ func (r *vmForVMClassPredicate) Generic(e event.GenericEvent) bool {
 func (r *vmForVMClassPredicate) matches(obj runtime.Object) (bool, error) {
 	obj = obj.DeepCopyObject() // do not mutate the original
 
-	vm, err := r.getVM(obj)
+	object, kind, err := r.getWatchedObj(obj)
 	if err != nil {
 		return false, err
 	}
 
+	// If the object is not VirutalMachine, then pass to the xxxToVMMapperFn
+	// for further processing.
+	if kind != "VirtualMachine" {
```

Comment:
Is there a reason this predicate is being used as a EventFilter (global predicate) instead of a predicate just on the specific VirtualMachine For()/Owns()/Watches() in the controller?

Reply:
This predicate mainly used for filter out Virtualmachine with a className , and the virtualmachineclass with controllerName that match the specific controller. 

There will be multiple controllers that will watch on virtualMachine in the feature, one example is we are going to use vmoperator to manage VMs on ESXi directly. The existing virtual machine controller will not work and there are a lot of limition on ESXi compare with vcenter.

Reply:
@yixingjia The way the event filter is setup, it gets called on every object the controller watches. In this case, it's `VirtualMachineClassBinding`, `ContentSourceBinding` and so on. Rather than excluding all the other types in the EventFilter predicate as done in the MR, we can get rid of these extra checks by just moving the predicate filter to the VirtualMachine objects only. The way to do that would be:
```golang
import (
  ...
  ctrlbuilder "sigs.k8s.io/controller-runtime/pkg/builder"
  ...
)

  builder := ctrl.NewControllerManagedBy(mgr).
  For(controlledType, ctrlbuilder.WithPredicates(kubeutil.VMForControllerPredicate(
			r.Client,
			// These events can be very verbose, so be careful to not log them
			// at too high of a log level.
			r.Logger.WithName("VMForControllerPredicate").V(8),
			vmClassControllerName,
			kubeutil.VMForControllerPredicateOptions{
				MatchIfVMClassNotFound:            true,
				MatchIfControllerNameFieldEmpty:   isDefaultVMClassController,
				MatchIfControllerNameFieldMissing: isDefaultVMClassController,
			}))).
```